apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/internal/SimpleResponseReader.kt

```diff
@@ -13,121 +13,64 @@ class SimpleResponseReader private constructor(
     private val variableValues: Map<String, Any?>,
     private val scalarTypeAdapters: ScalarTypeAdapters
 ) : ResponseReader {
+  private val responseRecordSetIterator = recordSet.iterator()
 
   constructor(
       recordSet: Map<String, Any?>,
       variables: Operation.Variables,
       scalarTypeAdapters: ScalarTypeAdapters
   ) : this(recordSet, variables.valueMap(), scalarTypeAdapters)
 
-  override fun readString(field: ResponseField): String? {
-    if (shouldSkip(field)) {
-      return null
-    }
+  override fun selectField(fields: Array<ResponseField>): Int {
+    while (true)
+      if (responseRecordSetIterator.hasNext()) {
+        val (nextFieldName, _) = responseRecordSetIterator.next()
+        val fieldIndex = fields.indexOfFirst { field -> field.responseName == nextFieldName }
```

Comment:
We could leverage the fact that field [order should be preserved](https://spec.graphql.org/June2018/#sec-Serialized-Map-Ordering) to remove the iteration here and have faster parsing performance than json. Maybe with a "slow" path for non-compliant server implementations.

Reply:
Even this is in 90% cases true, I wouldn't trust the order of fields we have in our models vs fields sent in query, we already do couple optimizations / merging fields, the order might changed and it's very fragile in terms of our codegen, any changes that impact fields order in our models everything will fall down as house of cards.

Reply:
> we already do couple optimizations / merging fields

Do you have an exemple for that?

We'd have to do some benchmarking but if we started the string matching on an index that increments as new fields are read and that'd work in 90% of the case, I think we'd save some time and be effectively faster than json (that'd be a cool marketing claim ;)). If tje 10% case fallbacks to iterating all the possible fields, everything would still work, albeit with performance on average the same as the one we have now?

Reply:
Inline fragments, in some conditions we merge fields. Also I'm not sure that we preserve 100% order when build IR or AST. If we want to do this we need to verify all over the place where we build IR and AST that we keep the original order of fields.

I would say let's first do perf measurement and see if it's a case, pretty sure we will get the same numbers as Moshi as it does the same field lookup.

Reply:
Damn inline fragments! But yea, I agree this all needs to be benchmarked.

Reply:
Actually after thinking we can do small optimization here, we can keep previously selected field index, then during next `selectField` call we can try to guess next field to select by `fields[previousSelectedFieldIndex + 1]` and if we fail fallback to lookup. Should optimize in 90%.

Reply:
Yes, exactly 👍

Reply:
I will address this in my next PR in `StreamResponseReader`.